test/integration/controller/controller_test.go

```diff
@@ -0,0 +1,783 @@
+	if errors.IsNotFound(err) {
+		return "", err
+	}
+	if err != nil || azVolume.Status.Detail == nil || azVolume.Status.Detail.ResponseObject == nil {
+		return "", status.Error(codes.Internal, "volume creation seems to have failed.")
+	}
+
+	return azVolume.Status.Detail.ResponseObject.VolumeID, nil
+}
+
+func checkReplicaCount(volumeName string, desiredNumReplica int) (bool, error) {
+	volumeID, err := getVolumeID(volumeName)
+	if len(volumeID) == 0 || err != nil {
+		return false, status.Errorf(codes.Internal, "failed to get volumeID with volumeName (%s)", volumeName)
+	}
+
+	diskName, err := azureutils.GetDiskNameFromAzureManagedDiskURI(volumeID)
+	if len(diskName) == 0 || err != nil {
+		return false, status.Errorf(codes.Internal, "failed to extract diskName from diskURI (%s)", volumeID)
+	}
+
+	labelSelector := fmt.Sprintf("%s=%s", azureutils.VolumeNameLabel, diskName)
+	azVAs, err := azVolumeClient.DiskV1alpha1().AzVolumeAttachments(namespace).List(context.Background(), metav1.ListOptions{LabelSelector: labelSelector})
+	if err != nil {
+		return false, status.Errorf(codes.Internal, "failed to get AzVolumeAttachment List for volumeName (%s): %v", diskName, err)
+	}
+
+	numReplicas := 0
+	for _, azVA := range azVAs.Items {
+		if azVA.Spec.RequestedRole == v1alpha1.ReplicaRole &&
+			azVA.Status.Detail != nil &&
+			azVA.Status.Detail.PublishContext != nil &&
+			azVA.Status.Detail.Role == v1alpha1.ReplicaRole {
+			numReplicas++
+		}
+	}
+	return numReplicas == desiredNumReplica, nil
+}
+
+func TestAzVolume(t *testing.T) {
+	if !*useDriverV2 {
+		klog.Infof("Skipping controller test... supported only for v2 driver")
+		return
+	}
+
+	err := testBootstrap()
+	assert.NoError(t, err)
+
+	tests := []struct {
+		testName            string
+		volumeName          string
+		capacityRange       *v1alpha1.CapacityRange
+		volumeCapabilities  []v1alpha1.VolumeCapability
+		parameters          map[string]string
+		secrets             map[string]string
+		volumeContentSource *v1alpha1.ContentVolumeSource
+		accessibilityReq    *v1alpha1.TopologyRequirement
+		testFunc            func(string)
+		cleanUpFunc         func(string)
+	}{
+		{
+			testName:   "create volume and check if finalizers have been properly populated.",
+			volumeName: "test-volume",
+			capacityRange: &v1alpha1.CapacityRange{
+				RequiredBytes: 1073741824,
+			},
+			volumeCapabilities: []v1alpha1.VolumeCapability{},
+			parameters: map[string]string{
+				"kind":      "managed",
+				"maxShares": "1",
+				"skuName":   "StandardSSD_LRS",
+			},
+			secrets:             map[string]string{},
+			volumeContentSource: nil,
+			accessibilityReq:    nil,
+			testFunc: func(volumeName string) {
+				azVolume, err := azVolumeClient.DiskV1alpha1().AzVolumes(namespace).Get(context.Background(), volumeName, metav1.GetOptions{})
+				if !assert.NoError(t, err) ||
+					!assert.NotNil(t, azVolume.Finalizers) {
```

Comment:
It looks weird to use `assert` functions in an `if` statement that does an early return. Oddly, the `assert` functions do return. Perhaps use the [`require`](https://pkg.go.dev/github.com/stretchr/testify/require) functions and wrap `testFunc` in `t.Run(...)`? The `require` functions use `t.FailNow(...)` which stops test execution.